vm/rust/src/execution.rs

```diff
@@ -324,39 +331,92 @@ fn set_l2_gas_limit(
+                    ..
+                }) => Ok(*all_resources),
+                _ => Err(anyhow::anyhow!("Unsupported Declare transaction version")),
+            },
+            AccountTransaction::DeployAccount(deploy_tx) => match &deploy_tx.tx {
+                DeployAccountTransaction::V3(DeployAccountTransactionV3 {
+                    resource_bounds: ValidResourceBounds::AllResources(all_resources),
+                    ..
+                }) => Ok(*all_resources),
+                _ => Err(anyhow::anyhow!(
+                    "Unsupported DeployAccount transaction version"
+                )),
+            },
+            AccountTransaction::Invoke(invoke_tx) => match &invoke_tx.tx {
+                InvokeTransaction::V3(InvokeTransactionV3 {
+                    resource_bounds: ValidResourceBounds::AllResources(all_resources),
+                    ..
+                }) => Ok(*all_resources),
+                _ => Err(anyhow::anyhow!("Unsupported Invoke transaction version")),
+            },
+        },
+        _ => Err(anyhow::anyhow!("Unsupported transaction type")),
+    }
+}
+
+/// Calculates the maximum amount of L2 gas that can be covered by the balance.
+fn calculate_max_l2_gas_covered<S>(
+    tx: &Transaction,
+    block_context: &blockifier::context::BlockContext,
+    state: &mut S,
+) -> Result<GasAmount, anyhow::Error>
+where
+    S: UpdatableState,
+{
+    // Extract initial resource bounds from the transaction.
+    let initial_resource_bounds = extract_resource_bounds(tx)?;
+
+    // Create resource bounds without L2 gas.
+    let resource_bounds_without_l2_gas = AllResourceBounds {
+        l2_gas: Default::default(),
+        ..initial_resource_bounds
+    };
+
+    // Calculate the maximum possible fee without L2 gas.
+    let max_fee_without_l2_gas =
+        ValidResourceBounds::AllResources(resource_bounds_without_l2_gas).max_possible_fee();
+
+    match tx {
+        Transaction::Account(account_transaction) => {
+            // Retrieve the fee token address.
+            let fee_token_address = block_context
+                .chain_info()
+                .fee_token_address(&account_transaction.fee_type());
+
+            // Get the balance of the fee token.
+            let balance = state
+                .get_fee_token_balance(account_transaction.sender_address(), fee_token_address)?;
+            let balance = (balance.1.to_biguint() << 128) + balance.0.to_biguint();
+
+            if balance > max_fee_without_l2_gas.0.into() {
+                // Calculate the maximum amount of L2 gas that can be bought with the balance.
+                let max_l2_gas = (balance - max_fee_without_l2_gas.0)
+                    / initial_resource_bounds
+                        .l2_gas
+                        .max_price_per_unit
+                        .0
+                        .max(1u64.into());
+                Ok(u64::try_from(max_l2_gas).unwrap_or(u64::MAX).into())
```

Comment:
I am unsure if we should go for u64::Max, we can be exposing ourselves to overflow errors during the execution. I am still not sure how could it happened but I remember some communication about it in the Shared Node meetings. Could we set `GasAmount::Max` here? Does it makes sense?

Reply:
GasAmount::Max is u64:MAX underneath

Reply:
Any reason why not use `GasAmount::Max` then?

Reply:
just type casting, let me change it